README.md
# Moneyhub Tech Test - Investments and Holdings

At Moneyhub we use microservices to partition and separate the concerns of the codebase. In this exercise we have given you an example `admin` service and some accompanying services to work with. In this case the admin service backs a front end admin tool allowing non-technical staff to interact with data.

A request for a new admin feature has been received

## Requirements

- An admin is able to generate a csv formatted report showing the values of all user holdings
    - The report should be sent to the `/export` route of the investments service
    - The investments service expects the report to be sent as csv text
    - The csv should contain a row for each holding matching the following headers
    |User|First Name|Last Name|Date|Holding|Value|
    - The holding should be the name of the holding account given by the financial-companies service
    - The holding value can be calculated by `investmentTotal * investmentPercentage`
- Ensure use of up to date packages and libraries (the service is known to use deprecated packages)
- Make effective use of git

We prefer:
- Functional code 
- Ramda.js (this is not a requirement but feel free to investigate)
- Unit testing

### Notes
All of you work should take place inside the `admin` microservice

For the purposes of this task we would assume there are sufficient security middleware, permissions access and PII safe protocols, you do not need to add additional security measures as part of this exercise.

You are free to use any packages that would help with this task

We're interested in how you break down the work and build your solution in a clean, reusable and testable manner rather than seeing a perfect example, try to only spend around *1-2 hours* working on it

## Deliverables
**Please make sure to update the readme with**:

- Your new routes
- How to run any additional scripts or tests you may have added
- Relating to the task please add answers to the following questions;
    1. How might you make this service more secure?
    2. How would you make this solution scale to millions of records?
    3. What else would you have liked to improve given more time?
  

On completion email a link to your repository to your contact at Moneyhub and ensure it is publicly accessible.

## Getting Started

Please clone this service and push it to your own github (or other) public repository

To develop against all the services each one will need to be started in each service run

```bash
npm start
or
npm run develop
```

The develop command will run nodemon allowing you to make changes without restarting

The services will try to use ports 8081, 8082 and 8083

Use Postman or any API tool of you choice to trigger your endpoints (this is how we will test your new route).

### Existing routes
We have provided a series of routes 

Investments - localhost:8081
- `/investments` get all investments
- `/investments/:id` get an investment record by id
- `/investments/export` expects a csv formatted text input as the body

Financial Companies - localhost:8082
- `/companies` get all companies details
- `/companies/:id` get company by id

Admin - localhost:8083
- `/investments/:id` get an investment record by id
- `/generate-report` (POST Method) creates a csv text of all users investments and POSTs it to the investments service

### Additional Scripts
Unit tests can be run from the admin directory using:
```bash
npm test
```

During development of a singular service the pm2 service can be used. pm2 runs and persists your node js app.
To install it: 
```bash
npm install pm2@latest -g
```
Then to start the dependent services:
```bash
cd investments 
pm2 start npm --name "investments-service" -- start

cd ../financial-companies
pm2 start npm --name "financial-companies-service" -- start
```

### Answers to Questions 
1. How might you make this service more secure?
    - The admin service should have some form of authentication middleware to prevent unauthorized persons from accessing its endpoints. Especially if there is a frontend UI pointing at it. Https is also a requirement for this
    - The other two services could be restricted to only allow access from the admin service providing a single point of entry. If hosted on docker or Kubernetes this can be done easily without ever exposing the investments or companies service to the outside world.
2. How would you make this solution scale to millions of records?

    There are multiple issues with the current implementation to allow scaling to millions of records. 
    - Firstly, the /generate-report endpoint waits for the csv to be generated before sending the response to the client. With millions of rows this would likely cause a request to time out. This could be solved in many different ways, such as: closing the request straight away with a 200 code and requiring the user to check back on progress through another API endpoint, or by using a websocket to provide a persistent connection with updates on progress.
    - Secondly, this could put significant load on the service while it processes the csv. It would be wise to employ load balancing and horizontal scaling to make sure other users can continue to use the service with no ill effects. 
    - Lastly, the investments service has a 10mb cap on data sent to it. Some form of streaming might need to be set up to allow the CSV to be sent to the investments service over time. An alternative to this would be to make the report smaller by offering the admin filters for things like the date range. 
3. What else would you have liked to improve given more time?
- Implement some more extensive unit tests
- Convert the project to ES6 import syntax to allow use of node-fetch v3 (v2 is still under support however)
- Make more use of the Ramda library (This is my first time using it and it won't be the last)
    - This includes replacing the nested for loops used to create the CSV  

### Additional notes
Please take my commenting style with a pinch of salt. It is not reflective of my real world usage, I wanted to use them to convey my thought process to the recruitment team around certain aspects of my code.

The way I brokedown this task should be clearly visible in the git history of this repo. If not there is a changelog with the rough order of events in the /admin directory.

admin/Changelog
### Changelog
- npm audit to remove vulnerabilities
- replaced depracated request module with node-fetch
    - node fetch v3 requires a workaround for projects using CommonJS
    - Given time the project could be updated to use ES6 imports
- updated investments endpoint to use node-fetch
- updated depracated res.send(status code) to res.sendStatus(status Code)
- Added helpers module for csv report logic to avoid bloat in index.js
- Split App and Server logic into two separate files to allow for Jest testing
- Implemented Unit testing of functions and API endpoints using Jest, fetch-mock-jest, and supertest
- Ran ESLint on code
- Updated README

admin/tests/admin.test.js
/* eslint-disable no-undef */
jest.mock("node-fetch", () => require("fetch-mock-jest").sandbox())
const fetchMock = require("node-fetch")
const request = require("supertest")
const {app} = require("../src/app")
const investmentsData = require("./MockData/investments.json")
const companiesData = require("./MockData/companies.json")
const config = require("config")

// Setup Mock (predicatable) responses from investments and financial-companies services

describe("Admin API test suite", () => {
  it("tests /investments/1 endpoint", async () => {
    fetchMock.mock(
      `${config.investmentsServiceUrl}/investments/1`,
      investmentsData[0],
    )
    request(app)
      .get("/investments/1")
      .then((response) => {
        expect(response.body).toEqual(investmentsData[0])
        expect(response.statusCode).toBe(200)
      })
  })

  it("tests /investments/2 endpoint when API fails", async () => {
    fetchMock.mock(`${config.investmentsServiceUrl}/investments/2`, 500)
    request(app)
      .get("/investments/2")
      .then((response) => {
        expect(response.statusCode).toBe(500)
      })
  })

  fetchMock.mock(
    `${config.investmentsServiceUrl}/investments`,
    investmentsData,
  )
  fetchMock.mock(
    `${config.financialCompaniesServiceUrl}/companies`,
    companiesData,
  )

  it("tests /generate-report endpoint returns successfully", async () => {
    fetchMock.mock(`${config.investmentsServiceUrl}/investments/export`, 204)

    request(app)
      .post("/generate-report")
      .then((response) => {
        expect(response.statusCode).toBe(200)
      })
  })

})


admin/tests/helpers.test.js
/* eslint-disable no-undef */
jest.mock("node-fetch", () => require("fetch-mock-jest").sandbox())
const fetchMock = require("node-fetch")
const investmentsData = require("./MockData/investments.json")
const companiesData = require("./MockData/companies.json")
const config = require("config")
const {makeCSV, fetchInvestmentData} = require("../src/helpers")

const expectedCSVString = `1,Billy,Bob,2020-01-01,The Small Investment Company,1400
2,Sheila,Aussie,2020-01-01,The Big Investment Company,10000
2,Sheila,Aussie,2020-01-01,The Small Investment Company,10000\n`

describe("CSV Helpers test suite", () => {
  it("makeCSV gives expected output given known input", () => {
    const csv = makeCSV(investmentsData, companiesData)
    expect(expect.stringMatching(csv)).toEqual(expect.stringMatching(expectedCSVString))
  })

  it("fetchInvestmentData gives expected csv given mock APIs", async () => {
    fetchMock.mock(
      `${config.investmentsServiceUrl}/investments`,
      investmentsData,
    )
    fetchMock.mock(
      `${config.financialCompaniesServiceUrl}/companies`,
      companiesData,
    )

    fetchInvestmentData().then(csv => {
      expect(expect.stringMatching(csv)).toEqual(expect.stringMatching(expectedCSVString))
    })
  })

})


admin/src/helpers.js
const R = require("ramda")
const config = require("config")
const fetch = require("node-fetch")

// Argument names containing type would be unecessary in TypeScript
// I could have put each row straight into the string but think the array provides a better debugging opportunity
const makeCSV = (investmentsArr, financialCompaniesArr) => {
  const csvArr = []
  investmentsArr.forEach(investment => {
    investment.holdings.forEach(holding => {
      const holdingCompany = R.find(R.propEq("id", holding.id))(financialCompaniesArr)
      csvArr.push([
        investment.userId,
        investment.firstName,
        investment.lastName,
        investment.date,
        holdingCompany.name,
        investment.investmentTotal * holding.investmentPercentage,
      ])
    })
  })

  let csvText = ""
  csvArr.forEach(row => {
    csvText += `${row.toString()}\n`
  })
  return csvText
}

const fetchInvestmentData = async () => {
  // This feels like an anti-pattern, I would appreciate feedback and an alternative solution
  const investments = await (await fetch(`${config.investmentsServiceUrl}/investments`)).json()
  const financialCompanies = await (await fetch(`${config.financialCompaniesServiceUrl}/companies`)).json()

  await Promise.all([investments, financialCompanies])
  return makeCSV(investments, financialCompanies)
}

module.exports.fetchInvestmentData = fetchInvestmentData
module.exports.makeCSV = makeCSV


admin/src/app.js
const express = require("express")
const bodyParser = require("body-parser")
const config = require("config")
// Using v2 of node-fetch, given time would like to convert to ES6 imports and v3
const fetch = require("node-fetch")
const helpers = require("./helpers")

const app = express()

app.use(bodyParser.json({limit: "10mb"}))

app.get("/investments/:id", (req, res) => {
  const {id} = req.params
  fetch(`${config.investmentsServiceUrl}/investments/${id}`)
    .then((response) => response.json())
    .then((body) => {
      res.send(body)
    })
    .catch((err) => {
      console.error(err)
      res.sendStatus(500)
    })
})

app.post("/generate-report", (req, res) => {
  // Fetching of data split into new module to keep index.js API endpoints readable
  helpers
    .fetchInvestmentData()
    .then((csv) => {
      // CSV is wrapped in a JSON object as investments service only parses JSON bodies
      // Modifying Investments service is out of scope of this task and in real world could be responsibility of another team
      fetch(`${config.investmentsServiceUrl}/investments/export`, {
        method: "post",
        body: JSON.stringify({report: csv}),
        headers: {"Content-Type": "application/json"},
      })
        .then((response) => {
          if (response.status !== 204)
            throw new Error("/investments/export didn't return 204")
          res.sendStatus(200)
        })
        .catch((err) => {
          console.error(err)
          res.sendStatus(500)
        })
    })
    .catch((err) => {
      console.error(err)
      res.sendStatus(500)
    })
})

module.exports.app = app
